test/distributed/pipelining/test_schedule_multiproc.py

```diff
@@ -932,6 +934,201 @@ def dw_runner():
                 ref_p = ref_submod.get_parameter(name)
                 torch.testing.assert_close(p.grad, ref_p.grad, rtol=1e-5, atol=4e-5)
 
+    @requires_nccl()
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "NCCL test requires 2+ GPUs")
+    @parametrize("ScheduleClass", [ScheduleGPipe, Schedule1F1B])
+    @parametrize("ModelClass", [MultiMLP, MultiMLPWithView])
+    @parametrize("LossType", ["mse", "mse_view"])
+    def test_schedule_memory_usage(self, ScheduleClass, ModelClass, LossType):
+        mod = ModelClass(d_hid, n_layers=self.world_size)
+        mod.to(self.device)
+
+        x = torch.randn(batch_size, d_hid, device=self.device)
+
+        # pre-run to allocate cuBLAS workspaces (see in test_fully_shard_memory.py)
+        mod(x).sum().backward()
+        torch.cuda.empty_cache()
+
+        x = torch.randn(batch_size, d_hid, device=self.device)
+        target = torch.randn(batch_size, d_hid, device=self.device)
+
+        mod.zero_grad(set_to_none=True)  # reset grads memory
+        base_mem = self._get_curr_active_memory()
+
+        if LossType == "mse":
+            loss_fn = torch.nn.MSELoss(reduction="sum")
+        elif LossType == "mse_view":
+            loss_fn = lambda x, y: torch.nn.functional.mse_loss(
+                x.view(x.size(0), -1), y.view(y.size(0), -1)
+            )
+
+        chunks = 4
+        x_mb = x.chunk(chunks)[0]
+
+        my_part = mod.get_submodule(f"layers.{self.rank}")
+        my_params_size = (
+            sum(p.numel() * p.element_size() for p in my_part.parameters()) / 1e6
+        )
+
+        split_spec = mod.split_spec if hasattr(mod, "split_spec") else None
+        pipe = pipeline(
+            mod,
+            mb_args=(x_mb,),
+            split_spec=split_spec,
+        )
+
+        stage = pipe.build_stage(
+            self.rank,
+            self.device,
+        )
+
+        schedule = ScheduleClass(stage, chunks, loss_fn=loss_fn)
+
+        current_mem = self._get_curr_active_memory()
+        self.assertLessEqual(
+            current_mem,
+            base_mem,
+            f"Rank {self.rank}: Creating the pipeline should not increase memory usage",
+        )
+
+        # Stage uses buffers internally for communications, we have to account for them
+        buffer_mem = x.numel() * x.element_size() / 1e6
+
+        if self.rank == 0:
+            schedule.step(x)
+            output_mem = 0
+
+            buffer_mem *= 1  # recv grads
+        elif self.rank == self.world_size - 1:
+            losses = []
+            output = schedule.step(target=target, losses=losses)
+
+            output_mem = (output.numel() * output.element_size()) / 1e6
+            buffer_mem *= 1  # recv acts
+        else:
+            schedule.step()
+            output_mem = 0
+
+        # these states are cleared at the beginning of each step, but not the end, so we need to clear them here
+        stage.clear_runtime_states()
+        gc.collect()
```

Comment:
Can we try running for several steps and remove the `gc.collect()`?

In prior experiments we found there were reference cycles due to not using `t.detach_()`. `gc.collect()` would detect any reference cycles, but we do not expect end users to call this every iteration because it is expensive. If we call it in the test we may be masking issues.